src/nupic/ntypes/Sdr.hpp

```diff
@@ -748,32 +795,201 @@ class SparseDistributedRepresentation : public Serializable
+        UInt index = 0;
+        for( ; index < callbacks.size(); index++ ) {
+            if( callbacks[index] == nullptr ) {
+                callbacks[index] = callback;
+                return index;
+            }
+        }
+        callbacks.push_back( callback );
+        return index;
+    };
+
+    /**
+     * Remove a previously registered callback.
+     *
+     * @param UInt Handle which was returned by addCallback when you registered
+     * your callback.
+     */
+    void removeCallback(UInt index) {
+        NTA_CHECK( index < callbacks.size() ) << "SDR::removeCallback, Invalid Handle!";
+        callbacks[index] = nullptr;
+    };
+
+    /**
+     * Add SDR to this classes proxy list.  When this SDR is destroyed, all
+     * proxies on the list will be cleaned up via a call to method
+     * proxy.deconstruct()
+     *
+     * @param proxy An SDR to register with this SDR for proper end-of-life
+     * handling.
+     */
+    void _addProxy(SparseDistributedRepresentation *proxy) {
+        proxies.push_back( proxy );
+    };
+
+    /**
+     * Remove an SDR from this classes proxy list.  Proxies must call this when
+     * they are destroyed so that this SDR can remove its pointers to the (now
+     * defunct) proxy.
+     *
+     * @param proxy An SDR to unregister with this SDR.
+     */
+    void _removeProxy(SparseDistributedRepresentation *proxy) {
+        for(auto it = proxies.begin(); it != proxies.end(); it++ ) {
+            if( *it == proxy ) {
+                proxies.erase( it );
+                return;
+            }
+        }
+        NTA_THROW << "SDR::_removeProxy, Proxy not found!";
     };
 };
 
 typedef SparseDistributedRepresentation SDR;
 
+/**
+ * SDR_Proxy class
+ *
+ * ### Description
+
+ * SDR_Proxy presents a view onto an SDR.
+ *      + Proxies have the same value as their source SDR, at all times and
+ *        automatically.
+ *      + SDR_Proxy is a subclass of SDR and be safely typecast to an SDR.
+ *      + Proxies can have different dimensions than their source SDR.
+ *      + Proxies are read only.
+ *
+ * SDR and SDR_Proxy classes tell each other when they are created and
+ * destroyed.  Proxies can be created and destroyed as needed.  Proxies will
+ * throw an exception if they are used after their source SDR has been
+ * destroyed.
+ *
+ * Example Usage:
+ *      // Convert SDR dimensions from (4 x 4) to (8 x 2)
+ *      SDR       A(    { 4, 4 })
+ *      SDR_Proxy B( A, { 8, 2 })
+ *      A.setSparse( {1, 1, 2}, {0, 1, 2}} )
+ *      auto sparse = B.getSparse()  ->  {{2, 2, 5}, {0, 1, 0}}
+ */
+class SDR_Proxy : public SDR
+{
+public:
+    /**
+     * Create an SDR_Proxy object.
+     *
+     * @param sdr Source SDR to make a view of.
+     *
+     * @param dimensions A list of dimension sizes, defining the shape of the
+     * SDR.  Optional, if not given then this Proxy will have the same
+     * dimensions as the given SDR.
+     */
+    SDR_Proxy(SDR &sdr)
```

Comment:
override to disable de/serialization of Proxy. Maybe only deserialization disabled, and allow serialization to SDR (call parent.save)?